lib/Doctrine/Common/Reflection/TypedNoDefaultReflectionProperty.php

```diff
@@ -0,0 +1,25 @@
+<?php
+
+namespace Doctrine\Common\Reflection;
+
+use ReflectionProperty;
+
+/**
+ * PHP Typed No Default Reflection Property - special override for typed properties without a default value.
+ */
+class TypedNoDefaultReflectionProperty extends ReflectionProperty
+{
+    /**
+     * {@inheritDoc}
+     *
+     * Checks that a typed property is initialized before accessing its value.
+     * This is neccessary to avoid PHP error "Error: Typed property must not be accessed before initialization".
+     * Should be used only for reflecting typed properties without a default value.
+     */
+    public function getValue($object = null)
+    {
+        $name = $this->getName();
```

Comment:
This should most certainly be cached locally (in the instance). Alternatively, using `$this->name` in the expression below would work too.

Yes, it's horrible, but this is really performance sensitive

Reply:
> This should most certainly be cached locally (in the instance). Alternatively, using `$this->name` in the expression below would work too.
> 
> Yes, it's horrible, but this is really performance sensitive

But what about `\Doctrine\Common\Reflection\RuntimePublicReflectionProperty`? It is using the same mechanism and a public property on an entity is still a much more common scenario than a typed property, caching should be implemented there too. Maybe a separate issue for this?

Reply:
And looking at the tests, adding caching would break the current behavior.
For example in https://github.com/doctrine/reflection/blob/7a2c95aa7c75613045de09985b5ddfdec05b9e18/tests/Doctrine/Tests/Common/Reflection/RuntimePublicReflectionPropertyTest.php#L13-L24

The RuntimePublicReflectionProperty would not follow the changes to the reflected object's properties, if it was caching the value on first access to the property.

Reply:
Also looking at 

https://github.com/doctrine/persistence/blob/779661acceefe09bd0e709b4cfe65ec0213f0d5c/lib/Doctrine/Persistence/Mapping/RuntimeReflectionService.php#L60-L74

The class introduced in this PR will be returned from this method too, and if we add caching only to this new class, it will still break the behavior of the `getAccessibleProperty` method, as the reflection property classes returned by it would not cache and track the reflected object changes for some properties and the opposite for other properties.